### ISSUE TITLE
Dell Latitude 7320 / 7420 with Intel i7-1185G7 throttling to 1800MHz

### ISSUE THREAD

**Issue by sebastianha** [intel/thermal_daemon#341]
### Preamble

I open this ticket as it gets clearer that not all recent Latitude models are affected in the same way, as I assumed earliert.

**This ticket is purely about the Dell Latitude 7320 (and 7420) with a i7-1185G7 (or similar) CPU.**

All the findings are present with Kernel 5.16.8-1-default and thermald master.

Reference tickets:
- https://github.com/intel/thermal_daemon/issues/293
- https://github.com/intel/thermal_daemon/issues/334

### Problem Description

Under high load (stress-ng -c 8) the CPU gets throttled very quickly down to 1800MHz. This would be fine if any thermal limits are hit but the Temperature does not exceed 55° and the fan does not even start spinning. Therefore, there should be options for the thermal management to keep the frequence higher at this point.

When running without thermald the frequency drops down to 400MHz, this problem was solved in an earlier version of thermald.

Running with thermald the following video shows typical behaviour (Please note, due to a bug in reading the cpu frequency it is shown in MHz not in GHz):

https://user-images.githubusercontent.com/420807/156352631-a7b13151-3df3-416a-b482-7b18f0f7e337.mp4

What we see:
- For a split second the frequency jumps up to >4.2GHz and the temperatures go up to >80°C.
- Then temperature goes down to ~55° and frequency drops to 1800MHz
- The fan did not start up at all
- Package power is limited 10W

Note: Fan sensor does not work, so we do not have any fan readings.

### Known things

I straced thermald and found out that I can reproduce the same behaviour by running
`echo 0 > /sys/devices/virtual/powercap/intel-rapl/intel-rapl:0/enabled`
instead of starting thermald: https://github.com/intel/thermal_daemon/issues/293#issuecomment-1018532302

I also tested setting power limits manually, they do not seem to have any effect.

Changing the Power mode in BIOS does not have any effect, "optimized" and "ultra performence" behave the same.

Available UUIDs are "UNKNOWN" but echoing the "adaptive" UUIDs has an effect:
```
> cat /sys/bus/acpi/devices/INTC1040:00/physical_node/uuids/available_uuids
UNKNOWN
```

There is a thing called "Official Ubuntu" from Dell, unfortunately I could not get hands on the one for a 7320 as I was not able to buy my notebook with this option. This is also the reason Dell refuses support for this problem in my case.

### Testing on Windows

When booting Windows and installing the Intel Dynamic Tuning Driver the CPU works as intended. Following https://github.com/intel/thermal_daemon/issues/334#issuecomment-1054355530 I ran the test and got the following readings:

[PwrData_2022-3-1-18-50-50_ultraperformance.csv](https://github.com/intel/thermal_daemon/files/8169214/PwrData_2022-3-1-18-50-50_ultraperformance.csv)

![Screenshot_20220302_095908](https://user-images.githubusercontent.com/420807/156353981-facf6c1a-74ac-404a-8423-c3ac3d508c1e.png)

We see here that for 60s the CPU is running at full speed, then the CPU gets regulated and therefore also the frequency. This is a known thing in the Windows world and the recommendation is to use a tool names "Throttlestop" to mitigate this. When locking "MMIO" via Throttlestop (values on throttlestop screenshot are not actual values) the CPU runs full speed for two minutes:

![mmio](https://user-images.githubusercontent.com/420807/156357505-3b859cdb-4386-4076-a3b1-f57bc287c7ae.png)

[PwrData_2022-3-1-18-55-29_ultraperformance_throttlestop_mmio-lock.csv](https://github.com/intel/thermal_daemon/files/8169231/PwrData_2022-3-1-18-55-29_ultraperformance_throttlestop_mmio-lock.csv)

![Screenshot_20220302_100324](https://user-images.githubusercontent.com/420807/156354370-1f5231b0-94cd-4f19-96be-6936168bcb8f.png)

### Current Thesis

My theory is that there is no control over the thermal management or something is wrong. This might be connected to the missing fan control but even without fan control the CPU should run at 95° before throttling. I think a "fallback thermal control" is limiting the CPU from damage which does not respect any sensor readings but directly limits the CPU to safe values.

The Windows tests are showing that the hardware is OK and that the performance can be achived via software. Unfortunately Dell is not willing to help, I already reached out to Dell support with no luck. There is also a large thread in the Dell forums: https://www.dell.com/community/Latitude/Latitude-5420-7420-7520-CPU-Throttling-Issue-on-Linux/td-p/7959019

### Goal

The goal should be that the CPU only gets throttled if all thermal regulations are maxed out. This means fan at full speed and CPU temperature at 95°C. Only at this point the frequency should drop to prevent overheating. I know from previous Latitude models that they are able to run full days on high temperatures and according to the specs I don't see why this one should not be able to do so.

### Closing

I really hope we can find a solution as there are many affected users out there which would like to "unleash" the full power of their devices. Thank you for alle the hard work on thermald!

**Comment by spandruvada** [comment]
We can't consider any result with "Throttlestop" as Windows performance. So in Windows also drops to 1800MHz after 60 seconds. Keep in mind CPU util 100% doesn't mean that all workloads have the same power requirement.

Let's do these experiments and see if any of this helps:


Disable Thermald

#systemctl disable thermald

Shutdown the system not reboot


1. Experiment 1


echo 200000000 > /sys/class/powercap/intel-rapl/intel-rapl:0/constraint_0_power_limit_uw
echo 0 > /sys/class/powercap/intel-rapl/intel-rapl\:0/enabled

echo 35000000 > /sys/class/powercap/intel-rapl-mmio/intel-rapl-mmio:0/constraint_0_power_limit_uw
echo 54000000 > /sys/class/powercap/intel-rapl-mmio/intel-rapl-mmio:0/constraint_1_power_limit_uw
echo 32000000 > /sys/class/powercap/intel-rapl/intel-rapl\:0/constraint_0_time_window_us


Power down and power up

2. Experiment 2

echo 200000000 > /sys/class/powercap/intel-rapl/intel-rapl:0/constraint_0_power_limit_uw
echo 0 > /sys/class/powercap/intel-rapl/intel-rapl\:0/enabled

echo 30000000 > /sys/class/powercap/intel-rapl-mmio/intel-rapl-mmio:0/constraint_0_power_limit_uw
echo 54000000 > /sys/class/powercap/intel-rapl-mmio/intel-rapl-mmio:0/constraint_1_power_limit_uw
echo 32000000 > /sys/class/powercap/intel-rapl/intel-rapl\:0/constraint_0_time_window_us

Power down and power up

3. Experiment 3

echo 200000000 > /sys/class/powercap/intel-rapl/intel-rapl:0/constraint_0_power_limit_uw
echo 0 > /sys/class/powercap/intel-rapl/intel-rapl\:0/enabled

echo 13500000 > /sys/class/powercap/intel-rapl-mmio/intel-rapl-mmio:0/constraint_0_power_limit_uw
echo 60000000 > /sys/class/powercap/intel-rapl-mmio/intel-rapl-mmio:0/constraint_1_power_limit_uw
echo 32000000 > /sys/class/powercap/intel-rapl/intel-rapl\:0/constraint_0_time_window_us

4. Experiment 4

echo 200000000 > /sys/class/powercap/intel-rapl/intel-rapl:0/constraint_0_power_limit_uw
echo 0 > /sys/class/powercap/intel-rapl/intel-rapl\:0/enabled

echo 28000000 > /sys/class/powercap/intel-rapl-mmio/intel-rapl-mmio:0/constraint_0_power_limit_uw
echo 60000000 > /sys/class/powercap/intel-rapl-mmio/intel-rapl-mmio:0/constraint_1_power_limit_uw
echo 32000000 > /sys/class/powercap/intel-rapl/intel-rapl\:0/constraint_0_time_window_us


5. Experiment 5

Experiment 1 settings +

echo sustained > /sys/bus/pci/devices/0000:00:04.0/workload_request/workload_type

6. Experiment 6
Experiment 2 +

echo semi_active > /sys/bus/pci/devices/0000:00:04.0/workload_request/workload_type

7. Experiment 7
Experiment 2 +

echo bursty > /sys/bus/pci/devices/0000:00:04.0/workload_request/workload_type

**Comment by sebastianha** [comment]
I just started doing the experiments but then noted something strange:

When I set
```
echo 0 > /sys/class/powercap/intel-rapl/intel-rapl:0/enabled
```
and then cat the value it is still "1":
```
cypod3:~ # echo 0 > /sys/class/powercap/intel-rapl/intel-rapl:0/enabled
cypod3:~ # cat /sys/class/powercap/intel-rapl/intel-rapl:0/enabled
1
cypod3:~ #
```

Also the constraint_0_time_window_us does not seem to be accepted by the system:
```
cypod3:~ # echo 32000000 > /sys/class/powercap/intel-rapl/intel-rapl:0/constraint_0_time_window_us
cypod3:~ # cat /sys/class/powercap/intel-rapl/intel-rapl:0/constraint_0_time_window_us
31981568
cypod3:~ #
```

It seems to be, that this is only a reading error as I get an effect when chaning the value. If set to "0" the minimum is 1800MHz. It still reads "1". But when then setting back to "1" it again drops down to 400MHz:

https://user-images.githubusercontent.com/420807/156892798-5cb5e193-af1b-4e59-84ff-647f255e4001.mp4

**Comment by sebastianha** [comment]
So, here are the results of the 7(+1) experiments. Summarized, I cannot see any changes in any experiment :(

### Preamble

All tests have been done directly after a fresh shutdown and power on. ThermalD is disabled and has not been started (current_uuid is "invalid"). These are the default values:

```
cypod3:~ # cat /sys/class/powercap/intel-rapl/intel-rapl:0/constraint_0_power_limit_uw
35000000
cypod3:~ # cat  /sys/class/powercap/intel-rapl/intel-rapl:0/enabled
1
cypod3:~ # cat /sys/class/powercap/intel-rapl-mmio/intel-rapl-mmio:0/constraint_0_power_limit_uw
15000000
cypod3:~ # cat /sys/class/powercap/intel-rapl-mmio/intel-rapl-mmio:0/constraint_1_power_limit_uw
18750000
cypod3:~ # cat /sys/class/powercap/intel-rapl/intel-rapl:0/constraint_0_time_window_us
27983872
cypod3:~ # cat /sys/bus/pci/devices/0000:00:04.0/workload_request/workload_type
none
```

**Durin all tests the fan was only barely spinning of completely off!**

I wrote a small script to generate graphs:

```
#!/usr/bin/env bash

SECONDS=0

echo "\"Seconds\",\"intel-rapl:0/constraint_0_power_limit_uw\",\"intel-rapl:0/enabled\",\"intel-rapl-mmio:0/constraint_0_power_limit_uw\",\"intel-rapl-mmio:0/constraint_1_power_limit_uw\",\"intel-rapl:0/constraint_0_time_window_us\",\"workload_type\",\"cpu_freq_0\",\"cpu_freq_1\",\"cpu_freq_2\",\"cpu_freq_3\",\"cpu_freq_4\",\"cpu_freq_5\",\"cpu_freq_6\",\"cpu_freq_7\",\"cpu_freq_avg\",\"package_temp\",\"cpu_temp_0\",\"cpu_temp_1\",\"cpu_temp_2\",\"cpu_temp_3\","

while true; do
        # Seconds since start
        echo -n "\"${SECONDS}\","

        # powercap settings
        echo -n "\"`cat /sys/class/powercap/intel-rapl/intel-rapl:0/constraint_0_power_limit_uw`\","
        echo -n "\"`cat /sys/class/powercap/intel-rapl/intel-rapl:0/enabled`\","
        echo -n "\"`cat /sys/class/powercap/intel-rapl-mmio/intel-rapl-mmio:0/constraint_0_power_limit_uw`\","
        echo -n "\"`cat /sys/class/powercap/intel-rapl-mmio/intel-rapl-mmio:0/constraint_1_power_limit_uw`\","
        echo -n "\"`cat /sys/class/powercap/intel-rapl/intel-rapl:0/constraint_0_time_window_us`\","
        echo -n "\"`cat /sys/bus/pci/devices/0000:00:04.0/workload_request/workload_type 2> /dev/null`\","

        # cpu frequency
        ALLFREQS=`cpupower -c all frequency-info | grep "asserted" | cut -d " " -f 6`
        MFREQ=0
        for i in $ALLFREQS; do
                if [[ $i != *.* ]]; then
                        i=`bc <<< "scale=3; $i / 1000"`
                fi
                echo -n "\"$i\","
                MFREQ=`bc <<< "scale=3; $MFREQ + $i"`
        done
        echo -n "\"`bc <<< "scale=3; $MFREQ / 8"`\","

        # temperatures
        TEMPS=`sensors | egrep "Package|Core" | cut -d "+" -f 2 | cut -d "°" -f 1`
        for i in $TEMPS; do
                echo -n "\"$i\","
        done

        echo ""
        sleep 1;
done
```

### Test Procedure

1. Boot up system and login into X11
2. Start s-tui in monitoring mode
3. Start turbostat and stat.sh
4. Verify system status (thermald off, uptime)
5. Follow commands from experiment
6. Start stress test in s-tui
7. While stress test is running check values with "cat"
8. After ~2 minutes stop turbostat and stat.sh, then stop stress test

Note that the graph is over the complete time. It starts before starting the stress test. So there could be some values/frequencies that are not relevant to the tests. See logs for exact time of starting. You can see it when the values given by the experient are in the log (here from second 26):

```
"25","200000000","1","15000000","18750000","27983872","none","2.39","2.27","2.04","2.35","2.34","2.20","1.90","2.38","2.233","54.0","50.0","48.0","47.0","54.0",
"26","200000000","1","35000000","54000000","31981568","none","2.95","2.64","2.60","1.71","2.76","2.80","2.56","2.38","2.550","51.0","51.0","45.0","48.0","47.0",
```

### Experiment 0 (behaviour without modifications)
Summary: Frequency drops down to 400MHz

![experiment_0](https://user-images.githubusercontent.com/420807/156895082-adb5a462-d995-4f49-be83-4edd44670ae6.png)

[experiment_0_stat.csv](https://github.com/intel/thermal_daemon/files/8191187/experiment_0_stat.csv)

[experiment_0_turbostat.out.tar.gz](https://github.com/intel/thermal_daemon/files/8191188/experiment_0_turbostat.out.tar.gz)

https://user-images.githubusercontent.com/420807/156895084-a61e4c6d-540b-4fbe-9f81-aa28b7a671e9.mp4

### Experiment 1
Summary: Frequency goes up for a very short amount of time, then drops to 1800MHz

![experiment_1](https://user-images.githubusercontent.com/420807/156895126-6ffe3bbe-89b0-479e-9353-0bf51b9388aa.png)

[experiment_1_stat.csv](https://github.com/intel/thermal_daemon/files/8191193/experiment_1_stat.csv)

[experiment_1_turbostat.out.tar.gz](https://github.com/intel/thermal_daemon/files/8191194/experiment_1_turbostat.out.tar.gz)

https://user-images.githubusercontent.com/420807/156895138-e0655a4e-2e37-45ef-af21-ce0a846c15a6.mp4

### Experiment 2
Summary: Frequency goes up for a very short amount of time, then drops to 1800MHz

![experiment_2](https://user-images.githubusercontent.com/420807/156895169-97a75083-b307-4202-ac42-65921c425d88.png)

[experiment_2_stat.csv](https://github.com/intel/thermal_daemon/files/8191196/experiment_2_stat.csv)

[experiment_2_turbostat.out.tar.gz](https://github.com/intel/thermal_daemon/files/8191197/experiment_2_turbostat.out.tar.gz)

https://user-images.githubusercontent.com/420807/156895185-211d7255-0d47-40f0-adf9-e402ec39ab37.mp4

### Experiment 3
Summary: Frequency goes up for a very short amount of time, then drops to 1800MHz

![experiment_3](https://user-images.githubusercontent.com/420807/156895199-4043de3f-eb81-4d68-8b4e-dc02c3ccf867.png)

[experiment_3_stat.csv](https://github.com/intel/thermal_daemon/files/8191198/experiment_3_stat.csv)

[experiment_3_turbostat.out.tar.gz](https://github.com/intel/thermal_daemon/files/8191199/experiment_3_turbostat.out.tar.gz)

https://user-images.githubusercontent.com/420807/156895211-9c6b3b4d-f896-4253-a28c-94475a14d400.mp4

### Experiment 4
Summary: Frequency goes up for a very short amount of time, then drops to 1800MHz

![experiment_4](https://user-images.githubusercontent.com/420807/156895242-a8656546-f1ff-42a0-836d-8c80dc335564.png)

[experiment_4_stat.csv](https://github.com/intel/thermal_daemon/files/8191200/experiment_4_stat.csv)

[experiment_4_turbostat.out.tar.gz](https://github.com/intel/thermal_daemon/files/8191201/experiment_4_turbostat.out.tar.gz)

https://user-images.githubusercontent.com/420807/156895249-cea9ba5a-2074-47c9-9a99-9adbe054f559.mp4

### Experiment 5
Summary: Frequency goes up for a very short amount of time, then drops to 1800MHz

![experiment_5](https://user-images.githubusercontent.com/420807/156895256-74ca2423-c857-430f-b677-655ee3848208.png)

[experiment_5_stat.csv](https://github.com/intel/thermal_daemon/files/8191202/experiment_5_stat.csv)

[experiment_5_turbostat.out.tar.gz](https://github.com/intel/thermal_daemon/files/8191203/experiment_5_turbostat.out.tar.gz)

https://user-images.githubusercontent.com/420807/156895265-6b6e128d-b6a4-4ce7-987f-509f8021ee39.mp4

### Experiment 6
Summary: Frequency goes up for a very short amount of time, then drops to 1800MHz

![experiment_6](https://user-images.githubusercontent.com/420807/156895273-1a7b0316-a5c0-4e35-ad60-71faead617df.png)

[experiment_6_stat.csv](https://github.com/intel/thermal_daemon/files/8191205/experiment_6_stat.csv)

[experiment_6_turbostat.out.tar.gz](https://github.com/intel/thermal_daemon/files/8191209/experiment_6_turbostat.out.tar.gz)

https://user-images.githubusercontent.com/420807/156895279-06820723-3c49-4aad-899c-90415ad6b203.mp4

### Experiment 7
Summary: Frequency goes up for a very short amount of time, then drops to 1800MHz

![experiment_7](https://user-images.githubusercontent.com/420807/156895284-6ef28125-5e27-4281-9f49-b9906d359de8.png)

[experiment_7_stat.csv](https://github.com/intel/thermal_daemon/files/8191207/experiment_7_stat.csv)

[experiment_7_turbostat.out.tar.gz](https://github.com/intel/thermal_daemon/files/8191208/experiment_7_turbostat.out.tar.gz)

https://user-images.githubusercontent.com/420807/156895295-943d38c3-8ff4-4d3b-8ccc-76e50451258e.mp4

### Summary

I think that something is still blocking the RAPL driver from doing its work. For me every experiment seems to behave the same, perhaps you see somthing in the log files or in the video that I am doing something wrong.

**Comment by spandruvada** [comment]
I need this system to analyze. I will request Dell, if they can loan me one.

**Comment by sebastianha** [comment]
I hope you get one. If I can do anything let me know.

If your request fails: would it help to boot up a live system on my notebook and give you SSH access?

Thanks in advance for all your efforts, I really appreciate it!

**Comment by spandruvada** [comment]
I see one possibility why Windows is different here. Trying that won't be easy without some kernel changes. So we have to enter a pause loop here.

**Comment by spamik** [comment]
I'll probably can compile kernel with patch if it helps...

**Comment by spandruvada** [comment]
Once I figure out details, I will give patch to try.

**Comment by sebastianha** [comment]
I should also be able to compile and boot a patched kernel.

**Comment by spandruvada** [comment]
Apply the attached patch to Linux kernel, build and install.
[0001-thermal-int340x-Add-new-attribute-for-policy-info.zip](https://github.com/intel/thermal_daemon/files/8219619/0001-thermal-int340x-Add-new-attribute-for-policy-info.zip)

When rebooted try these steps:
#systemctl stop thermald
#echo 7 > /sys/bus/platform/devices/INTC1040:00/enable_policy
Attach output of #dmesg
Then restart thermald
#thermald --no-daemon --loglevel=info --adaptive

**Comment by spamik** [comment]
I tried that but I got input/output error when I do echo command. In dmesg I got this:

[  135.400701] int3400 thermal INTC1040:00: enable_policy_store input: 7
[  135.402828] ACPI BIOS Error (bug): Could not resolve symbol [\_TZ.ETMD], AE_NOT_FOUND (20210930/psargs-330)
[  135.402834] ACPI Error: Aborting method \_SB.IETM._OSC due to previous error (AE_NOT_FOUND) (20210930/psparse-529)
[  135.402841] int3400 thermal INTC1040:00: enable_policy_store ret:-1

**Comment by sebastianha** [comment]
Hi,

first of all: **Seems that you did it!!!**

What I did:

1. Build new kernel module:

```
mkdir newkernel
cd newkernel/
make -C /usr/src/linux O=$(pwd) oldconfig
vi .config
make -j 3
```
then patch and run "make" again.

2. Load new module:

```
rmmod int3400_thermal
insmod ./int3400_thermal.ko
```

3. Test

- stop thermald
- "echo 7", this is the dmesg output:
```
[106550.332176] int3400 thermal INTC1040:00: enable_policy_store input: 7
[106550.335336] ACPI BIOS Error (bug): Could not resolve symbol [\_TZ.ETMD], AE_NOT_FOUND (20210930/psargs-330)
[106550.335363] ACPI Error: Aborting method \_SB.IETM._OSC due to previous error (AE_NOT_FOUND) (20210930/psparse-529)
[106550.335385] int3400 thermal INTC1040:00: enable_policy_store ret:-1
```
- start thermald

[thermald.log](https://github.com/intel/thermal_daemon/files/8222863/thermald.log)

- stress-test:

https://user-images.githubusercontent.com/420807/157650828-4769b775-6da4-4361-8fa3-559e7f806ee9.mp4

4. Result

Exactly like Windows temperatures go and also the frequencies for 60s. then it drops, but only to 2400MHz at 70°C

**Comment by spamik** [comment]
Hell, yeah, I can confirm. I stopped when the echo failed but after your post I've tried run stress test and I have 30W on CPU package and 99°C, runnint still on 3,9GHz. Wooohoooooo

**Comment by AdamantGarth** [comment]
Is it safe to try this patch on 5420, or is it 7320-specific?

**Comment by sebastianha** [comment]
I would say it is safe to try (I would do it), but to be safe only @spandruvada can tell.

**Comment by spandruvada** [comment]
This should be safe to run on 5420. But I don't think it will fix 5420.
For echo fail, can you try to write:
1, or 5 and see if it still fails?

**Comment by sebastianha** [comment]
I will test this evening and let you know.
I also will test combinations with BIOS performance settings (optimized vs ultraperformance)

**Comment by AdamantGarth** [comment]
If that's helpful, I get the same messages on 5420. As for performance - looks like I don't have this problem on 5420 anymore, because while compiling patched kernel my laptop's behavior was identical to @sebastianha's with the patch applied - 2.4GHz at 70°C. But now I rebooted with patched kernel and after 15 minutes of `s-tui` my laptop runs at ~2.7GHz, ~75°C. But to be sure I'll try to reboot back to regular kernel and run the test again.

**Comment by spandruvada** [comment]
@AdamantGarth Reboot may still may preserve the behavior. So you need to power off and boot.

**Comment by JoshuaPK** [comment]
@AdamantGarth it's worth noting that on my 5420, I've been able to "fix" the issue by removing the --adaptive option on the thermald startup.  Without --adaptive, mine behaves similarly to what you are describing here.  When I have some time tonight I'll try this kernel patch with the --adaptive switch in place to see if there is any improvement.

**Comment by sebastianha** [comment]
### Default Behaviour, BIOS Ultraperformance, with ThermalD

![stat_ultraperformance_thermald_default](https://user-images.githubusercontent.com/420807/157694315-8f219596-ea5b-4f6e-83d8-ec24b7fb81a2.png)

[stat_ultraperformance_thermald_default.csv](https://github.com/intel/thermal_daemon/files/8224633/stat_ultraperformance_thermald_default.csv)

As expected, dropping down to 1800MHz @50°C, some spikes

### Patch "7" Behaviour, BIOS Ultraperformance, with ThermalD

Procedure before test:
```
systemctl start thermald.service
systemctl stop thermald.service
rmmod int3400_thermal
insmod ./int3400_thermal.ko
echo 7 > /sys/bus/platform/devices/INTC1040:00/enable_policy
systemctl start thermald.service
```

![stat_ultraperformance_thermald_patch_7](https://user-images.githubusercontent.com/420807/157694531-1fc51442-f7c0-4edc-8f29-7966e079a023.png)

[stat_ultraperformance_thermald_patch_7.csv](https://github.com/intel/thermal_daemon/files/8224642/stat_ultraperformance_thermald_patch_7.csv)

As reported abode, power limiting after ~60s

### Patch "5" Behaviour, BIOS Ultraperformance, with ThermalD

Procedure before test:
```
systemctl start thermald.service
systemctl stop thermald.service
rmmod int3400_thermal
insmod ./int3400_thermal.ko
echo 5 > /sys/bus/platform/devices/INTC1040:00/enable_policy
systemctl start thermald.service
```

![stat_ultraperformance_thermald_patch_5](https://user-images.githubusercontent.com/420807/157694616-f5d8c696-2942-4dfe-89bb-4c6f3d14a656.png)

[stat_ultraperformance_thermald_patch_5.csv](https://github.com/intel/thermal_daemon/files/8224649/stat_ultraperformance_thermald_patch_5.csv)

Similar behaviour to "7". Clean shutdown between "7" and "5".

### Conclusion

The different power limit ramp up phase might be due to load before the test, as far as I know this is default thermald behaviour to incrase this slightly.

At the end it does not differ to me if I use BIOS setting "optimized" or "ultraperformance", The only difference is that subjectively the fan spins faster and louder.

Policy "5" or "7" does not matter for the results.

@spamik It would be interesting why your CPU stays at >90°C. Can you test with "s-tui" or "stress-ng" for better comparability?

**Comment by sebastianha** [comment]
@JoshuaPK  and @AdamantGarth : As the 5420 seems to be different than the 7320 I would like to keep this ticket clean and focused to solve one problem after another. Perhaps it would make sense to open also a ticket for the 5420, specifically.

**Comment by spamik** [comment]
I'll do s-tui test tomorrow when I finish system updates :-) I've run it before and it behaves exactly like this compilation test but I've been running it only for a while so I do longer test and let you know.

One funny, little offtopic, note. My friend has older 5491. It was running fine but he is regularly updating BIOS... now when I talk to him he find out that he has quite similar issue like our 7320. Under load power on s-tui throttle to 14W and so frequency goes down... It seems that not even Dell want to fix it but they are trying even to break older, until now good working, models :-)

**Comment by spandruvada** [comment]
@AdamantGarth Shutdown is fine. Removing power connector will not do more than that as there is battery,

**Comment by spandruvada** [comment]
@sebastianha Thanks for the tests. I will clean up the patch and submit to Linux. But since it too close to merge Window, it will may wait till 5.19.
I will start writing "5" to this entry in sysfs from thermald.

**Comment by spandruvada** [comment]
I suggest a dedicate thread for 5420.

**Comment by sebastianha** [comment]
> @sebastianha Thanks for the tests. I will clean up the patch and submit to Linux. But since it too close to merge Window, it will may wait till 5.19. I will start writing "5" to this entry in sysfs from thermald.

Thank you very much!

Is there somewhere an explanation what this exactly does? What is this "mysterious" new uuid and what is the difference between "5" and "7"?

**Comment by sebastianha** [comment]
I experimented with the power limitation after ~60s and I found that this must be one of the other temperature sensors kicking in. I can see in the thermald log:

```
[1646980616][INFO]Added zone 1 trip 4 clamp_valid 1 clamp 13500000
[1646980616][INFO]set cdev state index 11 state 13500000 wr:13500000
[1646980616][INFO]Set : 44500, 45050, 11, 13500000, 5000000
[1646980617][INFO]Erased  [1: 4 13500000
[1646980617][INFO]set cdev state index 11 state 35000000 wr:35000000
[1646980617][INFO]Set : 44500, 44050, 11, 35000000, 5000000
[1646980619][INFO]Added zone 1 trip 4 clamp_valid 1 clamp 13500000
[1646980619][INFO]set cdev state index 11 state 13500000 wr:13500000
[1646980619][INFO]Set : 44500, 45050, 11, 13500000, 5000000
[1646980634][INFO]Added zone 1 trip 3 clamp_valid 1 clamp 13500000
[1646980635][INFO]Erased  [1: 3 13500000
[1646980635][INFO]set cdev state index 11 state 13500000 wr:13500000
[1646980635][INFO]Set : 46000, 45050, 11, 13500000, 5000000
[1646980637][INFO]Added zone 1 trip 3 clamp_valid 1 clamp 13500000
```

**Comment by spamik** [comment]
One notice after new day: I've resumed system (suspended to disk yesterday) and I'm again on 1800MHz... I'm trying to do some tweaks to get it working again, hopefully I can get it working without rebooting...

**Comment by sebastianha** [comment]
I only do S0 suspend, this did not have any effect so far.

**Comment by spamik** [comment]
On S0 suspend my laptop survive 24 hours which is not much :-( And S3 is not supported on tigerlake CPU generation... :-/

**Comment by sebastianha** [comment]
As this is offtopic, I will send you a mail how to solve this. Short answer: Disable RAID in BIOS: https://bugzilla.kernel.org/show_bug.cgi?id=211879

**Comment by spamik** [comment]
After reboot it's working again. I've run s-tui stress test. Running over 10 minutes, behaviour is same from beginning: CPU freq in range 2,8-3GHz, temperatures over 60°C, cooling fan running on low speed. Yesterday it ras running differently - high speed on fan, higher frequency (4GHz at start, after minutes of loads drop to 3,2GHz), much higher temperature. But today I set cooling mode in BIOS from ultra performance to optimized so I believe that this behaviour change is in this setting and that if I return it to ultra performance I can get again higher.

![root@spm-ntb:-home-spm_001](https://user-images.githubusercontent.com/445877/157842576-f8ab0d67-fea8-4283-b561-d84759df70e5.png)

**Comment by spamik** [comment]
Update, after running my old fixing script it's working like yesterday (over 3GHz constantly) even on optimized mode:

#!/bin/bash
/etc/init.d/thermald stop
sleep 10
rmmod intel_rapl_msr
rmmod processor_thermal_device_pci_legacy
rmmod processor_thermal_device
rmmod processor_thermal_rapl
rmmod intel_rapl_common
rmmod intel_powerclamp
sleep 2
modprobe intel_powerclamp
modprobe intel_rapl_common
modprobe processor_thermal_rapl
modprobe processor_thermal_device
modprobe processor_thermal_device_pci_legacy
modprobe intel_rapl_msr
sleep 2
/etc/init.d/thermald start
echo "Done"

**Comment by sebastianha** [comment]
One interesting thing I found: I never see my CPU going >4300MHz. According to specs it should do 4800MHz. Is there something still blocking the full power?

**Comment by sebastianha** [comment]
Perhaps another notice when testing: It makes a different if the cpu is running in "performance" or in "powersave" mode.

In "powersafe" I noticed speeds at ~3000MHz and package power of ~16W. And of course with this setting the speed can maintained for a longer period of time  (>3 min) because the peripherals are not getting that hot.

Perhaps the task is to find a sweep spot for my work I am doing with the noteboot.

**Comment by Grtschnk** [comment]
> Apply the attached patch to Linux kernel, build and install. [0001-thermal-int340x-Add-new-attribute-for-policy-info.zip](https://github.com/intel/thermal_daemon/files/8219619/0001-thermal-int340x-Add-new-attribute-for-policy-info.zip)
> 
> When rebooted try these steps: #systemctl stop thermald #echo 7 > /sys/bus/platform/devices/INTC1040:00/enable_policy Attach output of #dmesg Then restart thermald #thermald --no-daemon --loglevel=info --adaptive

F*** yeah! It works, thank you @spandruvada !
Sustained 30W package power.

I will now keep an eye on the performance over the next week
I was also getting an input/output error for the echo command, similar to spamik, but does not break the functionality.

Also a big thank you to @sebastianha for investigating this issue :+1:

**Comment by Grtschnk** [comment]
After a reboot I am not able to reproduce it anymore, weird :-1:  :confused: 
I did change the bios settings to "Ultra" -> "Balanced" -> "Ultra". Will see if a made a mistake in the command sequence or if I accidentally changed some other BIOS setting.

**Comment by Oisangui** [comment]
Does this work for the 7420 model? I have the same problem!

**Comment by Grtschnk** [comment]
> Does this work for the 7420 model? I have the same problem!

Yep, I'm on a 7420 (CPU is the same as 7320).

**Comment by sebastianha** [comment]
I updated the title of this ticket for anyone searching for this problem

**Comment by n0rad** [comment]
archlinux on latitude 7420 here. 
I'm in quiet mode in the bios, not running thermald nor throttled daemons.

I built kernel `5.16.14` with the patch and played a bit. It's definitly better :+1: .

I confirm what was said, I have the same  input/output error when setting the flag

```
[  860.986898] int3400 thermal INTC1040:00: enable_policy_store input: 7
[  860.987241] ACPI BIOS Error (bug): Could not resolve symbol [\_TZ.ETMD], AE_NOT_FOUND (20210930/psargs-330)
[  860.987257] ACPI Error: Aborting method \_SB.IETM._OSC due to previous error (AE_NOT_FOUND) (20210930/psparse-529)
[  860.987271] int3400 thermal INTC1040:00: enable_policy_store ret:-1
```
But it's definitly doing something.

I also confirm that using `5` or `7` does not look to make a difference.

I can now see the CPU going up to `4.3Ghz 99°` on kernel compile, and most of the time `~3Ghz 80°`, instead of `1.8Ghz 55°`.

Now when I do a stress test (I also see that previous CPU status have impact on lauched test), the cpu goes to `~3.2Ghz ~99°`, but after ~45sec, it goes down to `~2.5Ghz ~80°` and will never go up again.

Starting `thermald` make the CPU back to  `~3.2Ghz ~99°` but only for 10sec.

I found that launching my previous startup script (https://github.com/erpalma/throttled/issues/255#issuecomment-903112374) finally make my CPU full blast `~3.4Ghz ~99°` without interruption. I don't know what is actually doing the trick in it.


**TLDR:** This startup script, with the kernel patch, makes my laptop finally fly :rocket: 

```bash
echo '7' | sudo tee /sys/bus/platform/devices/INTC1040:00/enable_policy

sudo rmmod intel_rapl_msr
sudo rmmod processor_thermal_device_pci_legacy
sudo rmmod processor_thermal_device
sudo rmmod processor_thermal_rapl
sudo rmmod intel_rapl_common
sudo rmmod intel_powerclamp

sudo modprobe intel_powerclamp
sudo modprobe intel_rapl_common
sudo modprobe processor_thermal_rapl
sudo modprobe processor_thermal_device
sudo modprobe processor_thermal_device_pci_legacy
sudo modprobe intel_rapl_msr
```

Definitely not the same laptop when starting flrefox with >1k tabs.

**Comment by spamik** [comment]
Just another question - as it seems that S0ix sleep won't be sufficient solution for me (Sebastian has best results of battery consumption what I have seen but even this is not suitable for my usage and it's far away from S3 sleep on my old laptop) I probably need to get working hibernation. But when I last hibernate laptop and then woke up it again, it has the same issue like without patch - locked to 1800MHz. I've trired removing and modprobing again mentioned modules even the one with patch but no change in behaviour - it start working only after reboot...

So is there anything what I can try to get same behaviour after resume from hibernation?

**Comment by spandruvada** [comment]
Remove the previous patch and apply the attached series.
[policy_0314_01.zip](https://github.com/intel/thermal_daemon/files/8248265/policy_0314_01.zip)

Build Thermal daemon from branch new_tgl_policy (checkout new_tgl_policy)
https://github.com/intel/thermal_daemon/tree/new_tgl_policy

Then test. Don't set sysfs manually. Thermald should do.
Also test hibernate case also.

**Comment by n0rad** [comment]
built the new kernel and thermald.

```bash
sudo ./thermald --no-daemon --loglevel=info --adaptive | grep "Set Default"

[1647295281][INFO]Set Default UUID: 42A441D6-AE6A-462b-A84B-4A8CE79027D3
```

Running bench get to `~4.3Ghz 99°` for around 15sec and then getting down to `~2.8Ghz 72°` forever. Triggering modules reload make it able to stay at `XX 99°`
I did not tested hibernation.

Using the laptop today for work was much more responsive, and the CPU is giving as much power as possible on demand till getting to  `99°` (around 2.9Ghz on the long run).
but I noticed the fan is not giving everything it can, cause sometimes when I suspend my laptop, fan get full speed (I don't know why), and it's turning much faster than on demand :thinking:

**Comment by sebastianha** [comment]
The throttling is caused by other sensors not the CPU. You see this also in thermald log. Unfortunately, Dell does not a great job in cooling. If you stop thermald this safety measure is not kicking in.

However, I don't know if the others hardware might work with higher temperatures. For me I am getting throttled as soon as one of the sensors reaches ~47°C, see graph:

![signal-2022-03-14-124239_001](https://user-images.githubusercontent.com/420807/158332390-9630e6e0-c70a-4d35-8fa2-381b9f4ee272.png)

So it depends on cooling and the ambient temperature how long it takes until you get throttled. The CPU actually is handled well and only is throttled down because the power limit applies to "save" the rest of the hardware.

**Comment by spamik** [comment]
I've tried new kernel patches and thermald from tgl_new_policy branch. I've noticed that thermald has to run with --adaptive option (without it it's not working) but yes, it's working OK (same behaviour for me like previous patch, no difference) without echoing policy.

Unfortunately hibernation still not working. After resume I'm still stuck on 1800MHz no matter which kernel modules I reload, if I restart thermald and so on.

**Comment by spandruvada** [comment]
Does rmmod and again modprobe of int3400_thermal helps after hibernate?

**Comment by spamik** [comment]
I've tried it among with int3403_thermal which I have also loaded but no change. But I did it with other modules above. Should I try to unload/load just this module without other or it didn't change anything?

**Comment by spandruvada** [comment]
hibernate calls some special calls into firmware, May be that is not a tested path.

**Comment by spamik** [comment]
Can help something like unloading module before hibernation?

**Comment by sebastianha** [comment]
Perhaps a silly question but: Does this work on Windows? Do you get full power there after hibernation? I can't test this well because I removed Windows from my ssd and am only booting from USB but this might be worth a try if someone has "real" multiboot.

**Comment by spamik** [comment]
To be honest, I'm not sure, I've booted it once just to extract product key and then I removed it... I believe that on Win it's working as AFAIK Win in default settings do hybrid sleep (after some time in suspend to RAM it hibernate itself) so probably there will be a lot of users suffering from same issue.

**Comment by spamik** [comment]
Ok, so I compiled kernel with added this patch, turned off laptop, boot, reload modules and start thermald with --adaptive option. Working good. Suspend to disk, resume, issue persist (I've tried again to reload modules, restart thermald, still stuck on 1800MHz). Regarding log, after first clean boot I got this:

```
spm-ntb /home/spm # dmesg |grep OSC
[    1.503928] ACPI: \_SB_.PR00: _OSC native thermal LVT Acked
[    2.153053] acpi PNP0A08:00: _OSC: OS supports [ExtendedConfig ASPM ClockPM Segments MSI HPX-Type3]
[    2.158472] acpi PNP0A08:00: _OSC: platform does not support [AER]
[    2.168552] acpi PNP0A08:00: _OSC: OS now controls [PCIeHotplug SHPCHotplug PME PCIeCapability LTR]
[  106.167135] ACPI Error: Aborting method \_SB.IETM._OSC due to previous error (AE_NOT_FOUND) (20210930/psparse-529)
```
and after resume these lines (again grepping OSC) appears:

```
[  165.520525] ACPI Error: Aborting method \_SB.IETM._OSC due to previous error (AE_NOT_FOUND) (20210930/psparse-529)
[  165.520530] Reset _OSC cap to 0
[  165.982900] ACPI Error: Aborting method \_SB.IETM._OSC due to previous error (AE_NOT_FOUND) (20210930/psparse-529)
[  165.982904] Restore _OSC cap to 0x2
[  234.991731] ACPI Error: Aborting method \_SB.IETM._OSC due to previous error (AE_NOT_FOUND) (20210930/psparse-529)

```

**Comment by spandruvada** [comment]
It appears you only suspended for less than a second. Or your suspend to disk failed and there was immediate wake.
So hibernate (aka S4) was not successful.

**Comment by spamik** [comment]
It's weird because from user perspecitve it was working properly - I click on suspend to disk, it takes several seconds to do it, after it laptop correctly turned off, then I powered it on, see grub, entered password for LUKS and after another few seconds I got my desktop environment with running browser as I leave it before.

**Comment by spandruvada** [comment]
Apply this patch 
[init_osc.zip](https://github.com/intel/thermal_daemon/files/8274808/init_osc.zip)

And after resume just rmmod int3400_thermal and modprobe again.

**Comment by spamik** [comment]
Unfortunately still the same :-( I applied this patch on the top of the all before if it make some difference.

**Comment by spandruvada** [comment]
Not sure what we can do more. This probably works same in Windows.

**Comment by spamik** [comment]
I can try to ask on dell comunity forum if someone with Windows can confirm this.

**Comment by sebastianha** [comment]
I will give it a try, but I am not sure if this test is valid as I am booting from an USB-C disk (SanDisk Extreme 1 TB Portable SSD)

**Comment by sebastianha** [comment]
Ok, hibernate is not supported when booting from USB. I tried to enable hibernate but got an error, that "Windows to go" has blocked it. So I cannot test, sorry.

**Comment by spandruvada** [comment]
Can you confirm on Linux that without hibernate the patches for suspend/resume works fine with normal, lid close or sleep.
https://github.com/intel/thermal_daemon/files/8263165/0001-thermal-int340x-Add-suspend-callback.zip
On top of previous patches https://github.com/intel/thermal_daemon/files/8248265/policy_0314_01.zip

Thermald can still work without throttle. I want to see if I can submit the suspend/resume callback patch to upstream or not as it still do right things.

**Comment by spamik** [comment]
Yes, I have applied all these patches and without hibernation it's working OK - clean boot full performace, with suspending to S0ix and resuming it's also working OK.

**Comment by spandruvada** [comment]
Changes released Thermal 2.4.9. The kernel patches in linux-next queued for 5.18

**Comment by Oisangui** [comment]
This led me to build my first kernel. It worked! Thank you!!

**Comment by sebastianha** [comment]
I would consider this issues as solved?! Any vote against it?

(for hibernation I would ask to create a new ticket)

**Comment by spandruvada** [comment]
The kernel patch is applied to upstream tree:

commit c7ff29763989bd09c433f73fae3c1e1c15d9cda4
Author: Srinivas Pandruvada <srinivas.pandruvada@linux.intel.com>
Date:   Mon Mar 14 15:09:37 2022 -0700

    thermal: int340x: Update OS policy capability handshake
    
    Update the firmware with OS supported policies mask, so that firmware can
    relinquish its internal controls. Without this update several Tiger Lake
    laptops gets performance limited with in few seconds of executing in
    turbo region.

**Comment by spandruvada** [comment]
The issue with _OSC error is BIOS related issue, In spite of error the feature still works. This can't be solved by thermald.
Without calling _OSC then we will have the frequency stuck issue.
So please close this issue.

**Comment by sebastianha** [comment]
FYI: the patch has been released with Kernel 5.18-rc1

**Comment by PhilipGB** [comment]
> 

Just tested this with Ubuntu 22.04 Beta and it happily sits at 2.6ghz using stress. Shame the official kernel will be 5.15 though maybe Canonical will backport the fix

https://bugs.launchpad.net/ubuntu/+source/thermald/+bug/1958328/comments/3

**Comment by sebastianha** [comment]
Waiting on openSUSE Tumbleweed for the new Kernel, thermald has already been updated to 2.4.9. Will report when Kernel 5.18 has been released.

**Comment by ben75020** [comment]
Upgraded ubuntu to 22.04 with latest kernel 5.18-rc1, s-tui now offers stable values of 3.1 ghz while CPU being stressed. 

Don't get fooled by latest s-tui embeded into Ubuntu 22.04 : due to a [python library bug,](https://github.com/amanusk/s-tui/issues/186) units are shown divided by 1000. 

Could someone re-explain what was the problem in simple words ?

**Comment by sebastianha** [comment]
Here is an explanation of the change: https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/?h=next-20220318&id=c7ff29763989bd09c433f73fae3c1e1c15d9cda4

In short (as far as I understand) the old way of setting a thermal policy is no longer supported by the latest CPUs and the new method had to be implemented in the kernel and to be used by thermald.

**Comment by gm42** [comment]
> Just tested 5.18rc1 with Ubuntu 22.04 Beta and it happily sits at 2.6ghz using stress. Shame the official kernel will be 5.15 though maybe Canonical will backport the fix

The fix is already backported for `jammy` (22.04): https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1966089
It mentions kernel version `5.15.0-25.25`.

I have an `i7-1165G7` however `thermald` reports:
```
27 CPUID levels; family:model:stepping 0x6:8c:1 (6:140:1)
[/sys/devices/platform/thinkpad_acpi/dytc_lapmode] present: Thermald can't run on this platform
Unsupported cpu model or platform
```
So I can't use `thermald` on this machine.
For reference, I did not blacklist the `intel_rapl_msr` module and with SpeedStep disabled from BIOS I see CPU staying at a fixed 1.2GHz while with SpeedStep enabled and a `performance` CPU governor I can see reasonable behavior happening: CPU will ramp up to 4.1GHz and then settle around 3.1GHz during an intensive build job.

The fan behavior is also not optimal so I use [thinkfan](http://manpages.ubuntu.com/manpages/bionic/man1/thinkfan.1.html) to avoid an always-spinning fan at temperatures of ~49C.

**Comment by gm-vm** [comment]
@spandruvada I have a Latitude 7520, which also has a i7-1185G7, and I think commit c7ff29763989 ("thermal: int340x: Update OS policy capability handshake") did not fix all the issues. Not for this laptop at least.

There seems to be a known workaround that involves reloading some kernel modules and after trying it I can confidently say that that somehow solves the problem.

Debugging a bit more I found that what seems to unlock the full speed of the CPU is `proc_thermal_rapl_remove()`. All I need to unlock the full speed of the CPU is `rmmod processor_thermal_device_pci_legacy`, I do not even need to modprobe it again. 

For now I'm reloading the module on system startup, but this obviously not great and it may break in future.

Should I report this issue elsewhere?

**Comment by sebastianha** [comment]
That's really interesting that these similar sounding Dell models are so different in design/BIOS/software.

**Comment by gm-vm** [comment]
OK, I think I figured out something.

When `processor_thermal_device_pci_legacy` is unloaded, what ultimately bumps the speed of the CPU is `rapl_remove_package`. From what I can see, this function removes all the power limits (or maybe it sets the hardware's default?).

Changing `constraint_*_power_limit_uw` before unloading the module allows to bump the CPU clock to the expected value. I do not know why I thought this was not working, maybe I tried before the kernel fix or who knows what I did.

What ultimately confused me was that the following prints the same stuff before unloading the module and after reloading it:
```sh
$ grep . /sys/devices/virtual/powercap/intel-rapl-mmio/intel-rapl-mmio:0/constraint_*_power_limit_uw
/sys/devices/virtual/powercap/intel-rapl-mmio/intel-rapl-mmio:0/constraint_0_power_limit_uw:15000000
/sys/devices/virtual/powercap/intel-rapl-mmio/intel-rapl-mmio:0/constraint_1_power_limit_uw:18750000
```

However, I think that for the second run the values printed are not the real ones. Moreover, changes to `constraint_*_power_limit_uw` after reloading the module seem to have no effect, so clearly there is something wrong.

In short, change `constraint_*_power_limit_uw` without unloading the module.

**Comment by sebastianha** [comment]
WARNING: DO NOT UPDATE TO LATEST BIOS (1.16.2)

I just bootet Windows and it did this automatically :(

Now I am stuck at 1,6GHz and and it looks like the new module/thermald has no effect anymore.

**Comment by sebastianha** [comment]
(Deleted last post, where I though that the patch does no longer work with latest BIOS. Turns out was my fault, ignore it. Everything is fine!)

**Comment by 0hax** [comment]
I'm running latest kernel 5.18.0-3-amd64 on Debian with thermald 2.5.0.
I have also upgraded the BIOS firmware to 1.17.2.
And I'm still seeing my CPU at 400Mhz with or without thermald running.

This is for a Dell latitude 7420 with this CPU:
```
processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 140
model name      : 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
stepping        : 1
microcode       : 0xa4
cpu MHz         : 1800.000
cache size      : 12288 KB
physical id     : 0
siblings        : 8
core id         : 0
cpu cores       : 4
apicid          : 0
initial apicid  : 0
fpu             : yes
fpu_exception   : yes
cpuid level     : 27
wp              : yes
```

Does someone have the same issue on this version?

**Comment by Sprungles** [comment]
I found a temporary fix. The only thing you DO NOT DO in this article is disable Dell BIOS fan control. ( https://mijorus.medium.com/thermal-tweaks-on-dell-laptops-running-linux-769c8c80022e )This will allow you to control the fans via your OS. I have tested this on Linux and have only approximately 10% CPU usage total on idle. This is significantly better than it was before. Until a custom bios is made, we may be out of luck on a hardware level.

**Comment by sebastianha** [comment]
What do you want to fix? With the current Kernel and thermald my system runs perfectly.

I set the thermal profile to Quiet which works very well for me: `smbios-thermal-ctl --set-thermal-mode=Quiet`

One thing I noticed: The fan is not only responsible to cool the CPU but also the GPU and NVME. Especially when the NVME gets hot it starts spinning. Unfortunately relatively early at low temperatures. But I can live with it.

**Comment by Sprungles** [comment]
I ended up using a utility known as i8kmon to increase the fan control
after do as you did with thermald. It is now working well on the dell
latitude e7450. I hope that this helps, as I spent many hours getting this
to work.
